Modify text rendering inside notes to add styling

dmsnell opened the pull request:
<del>Resolves automattic/wp-calypso#2044</del> (resolved via API)

<del>A long-standing issue has been the lack of spacing between paragraphs of
text. Every line-break has received a `<br />` tag to separate them but
this left paragraphs flowing without break.</del>

This patch is a hack that adds on some primitive styling in order to get
"the best bang for the buck" until a proper refactor of the text
processing occurs.

Namely:
 - Detect lists and make them real lists with `<ul>` and `<li>`
 - Detect emphasized headers like `Next Week: do something`
 - Split paragraphs of text vertically

The new `toBlocks` reducer does the real work here and takes a fairly
naive approach. Certain bugs exist but I believe that the work in this
patch will relieve much more frustration than it possibly introduces.

@drw158 can you review the use of the bullet? I know there are some styling things I just blasted in without knowing how to do them. also, I seem to remember that we may have intentionally prevented the use of `<li>` so if the bullet isn't right we can change that.

**Testing**
Boot up with `npm start` and do a visual inspection of the notes. Look for inconsistencies in the rendering or blatant problems.

roundhill commented:
Fixed on the api side by @dmsnell 👍

dmsnell commented:
@roundhill would still like some feedback here for bullet lists and emphasized "header" text

think it's worth using?

roundhill commented:
> @roundhill would still like some feedback here for bullet lists and emphasized "header" text
> 
> think it's worth using?

Oh! Sorry for prematurely closing this. I think it might be worth adding these additional formatting niceties. Any thoughts @drw158?

drw158 commented on src/boot/stylesheets/main.scss:
```diff
@@ -690,12 +690,21 @@ body {
 
+		.wpnc__body-list {
+			list-style: disc;
+			margin-left: 2em;
```
I would do this instead of margin-left

`padding: 0 16px;`

That'll line things up on the sides.

![screen shot 2017-05-23 at 10 48 13 am](https://cloud.githubusercontent.com/assets/618551/26363188/f1ce6a2a-3fa5-11e7-9502-af2266d834c1.png)

drw158 commented:
The meeting note notification looks great:

![screen shot 2017-05-23 at 10 48 13 am](https://cloud.githubusercontent.com/assets/618551/26363230/0edb4c32-3fa6-11e7-94dd-0cd51519c416.png)

On an older comment, for some reason the list tags are not used, and the bullet is being inserted directly.

dmsnell commented:
@drw158 thanks for the CSS suggestions. I added a new chunk for rendering inline code and I took direct inspiration from how Slack does it. you might want to review that.

@roundhill removed the paragraph detection and added todo lists as well!

ready for another review!

gwwar commented on src/templates/functions.jsx:
```diff
@@ -2,16 +2,96 @@ import React from 'react';
+ * @param {String} text input list of blocks
+ * @returns {String} marked-up text
+ */
+const toBlocks = text =>
```
Is text already stripped of html by this point? 

Overall this looks ok so far for a quick and dirty replace. I'd recommend adding unit tests for these cases so refactoring is pretty easy.

dmsnell replied:
actually the comment is misleading, it's an HTML string so markup is already inside of it because of other processing in the app. the note data comes from the API without any HTML though.

drw158 commented on src/boot/stylesheets/main.scss:
```diff
@@ -690,12 +690,41 @@ body {
+		}
+
     	blockquote {
     		margin: 0 $padding-medium;
```
Right now blockquotes look weird because they do not have a bottom margin. I think before the last commit, they were wrapped in a `p` maybe.

Change margin to this and it should look better:

`margin: 0 $padding-medium $padding-medium;`

dmsnell replied:
Updated margin in 08a122e

drw158 commented:
Mentions seem to be broken, cutting off a lot of text:

![screen shot 2017-05-24 at 10 50 46 am](https://cloud.githubusercontent.com/assets/618551/26412603/ec2c0be4-406e-11e7-90d1-057865677c85.png)
![screen shot 2017-05-24 at 10 50 20 am](https://cloud.githubusercontent.com/assets/618551/26412604/ee275a84-406e-11e7-9c1f-c0d86ab639bf.png)

drw158 commented on src/boot/stylesheets/main.scss:
```diff
@@ -690,12 +690,41 @@ body {
     		}
+
+			code {
+				font: $code;
```
change to `font-family` to solve the invalid property error.

dmsnell replied:
adopted in 08a122e

drw158 commented on src/boot/stylesheets/main.scss:
```diff
@@ -690,12 +690,41 @@ body {
+				background-color: $code-background;
+				border: 1px solid $code-border;
+				border-radius: 3px;
+				padding-left: 2px;
```
Might as well write this as:

`padding: 0 2px;`

dmsnell replied:
compressed padding in 08a122e

drw158 commented on src/boot/stylesheets/main.scss:
```diff
@@ -690,12 +690,41 @@ body {
+			code {
+				font: $code;
+				font-size: 90%;
+				color: $code-color;
```
Nice, we don't really have a style for `code` in Calypso. Looks good, but with some color changes. First, I wouldn't use variables since this is a one-off deal and we try to keep the colors.scss file pretty clean.

Change the colors to this:

```
color: darken( $gray, 20% );
background: $gray-light;
border: 1px solid lighten( $gray, 20% );
```

dmsnell replied:
inlined in 08a122e

dmsnell commented:
> Mentions seem to be broken, cutting off a lot of text:

@drw158 is there a way I can try to reproduce this. that seems odd and I want to make sure this code isn't doing something really funny but I'm not seeing the behavior/don't know if I'm seeing it

dmsnell commented:
Good call @drw158! in 798f875f5ba64a58c9e9f0e3fa926b994f2efa0f I fixed this behavior. it was a bug due to my misunderstanding of `String.split()` and has been fixed. I checked for other places this could have creeped in and I'm not using any other instances of that incorrectly.

Actually it was related to having a header-style paragraph with URLs (`:`s) in and not related to the mention.

gwwar commented on src/templates/functions.jsx:
```diff
@@ -22,28 +102,23 @@ export function p(html, className) {
         return (
             <div
                 className={className}
-                key={somewhatRandomKey}
```
😄 `somewhatPsuedoRandomKey`

gwwar commented on src/templates/functions.jsx:
```diff
@@ -2,16 +2,96 @@ import React from 'react';
+            : raw;
+
+        // inline `code` snippets
+        const coded = defined.replace(/`([^`]+)`/, '<code>$1</code>');
```
I noticed a minor corner case with a `-` in a code block. I think we can probably ignore this for now, but might be worth considering constructing a proper tokenizer later if we need to handle more difficult cases in the future.
```
`- foo/bar`
```
<img width="346" alt="screen shot 2017-05-31 at 3 39 12 pm" src="https://cloud.githubusercontent.com/assets/1270189/26657218/6aa6876c-4617-11e7-83fe-2851b026dc43.png">

dmsnell replied:
@gwwar I absolutely want to do this properly. it's basically only because I'm tired of seeing such unstyled content that I did this as a "meh, it's much better than it was"

I think you are right though in that inline code should have precedence over a list and I'll make this change.

having had such major issues in the past held me back from making some of these smaller improvements. when we "do it properly" (and why I hesitate from just doing that now) I want to make a drastic change in how we do all of the transformation. right now we pump API responses directly into the app and transform the data on every render. the approach I took in [dmsnell/wrnc](https://github.com/dmsnell/wrnc) was to [transform the API data when it comes into the system](https://github.com/dmsnell/wrnc/blob/58fe073c1c0b196ee46d02eb8dc178fbdc3c4227/src/api-poller/from-api.js) then we never have to repeat the transformation again. much faster, much saner. (data models, who would'a thunk it?)

dmsnell replied:
@gwwar I can't reproduce. are we sure that isn't actually a bullet followed by inline code? (inline code already has precedence over lists)

can you share the underlying text causing this issue?

gwwar replied:
It's a code block over multiple lines eg
```
<code>
- foo
- bar
- baz
</code>
```

See p4TIVU-6vp-p2#comment-7105

dmsnell commented on src/boot/stylesheets/main.scss:
```diff
@@ -690,14 +690,53 @@ body {
+					border: none;
+					background: none;
+				}
+			}
```
@drw158 I added this section here because I was getting funny styling with `<code><pre>`

<img width="189" alt="screen shot 2017-05-31 at 9 58 42 pm" src="https://cloud.githubusercontent.com/assets/5431237/26663446/495689d4-4650-11e7-8ef3-c987463c337f.png">

I would be very grateful if you reviewed this here CSS…

dmsnell commented:
Is it worth not pushing this that the mobile clients don't support it?

drw158 commented:
I think it's worth pushing even if mobile clients don't support it.
